Show public role entry in sys.server_principals

aswiniip opened the pull request:
### Description
Currently in Babelfish we don't show any `public` entry. We should show the public role as well. To achieve this I have added a `public` entry in the sys.server_principals by making changes in the ownership.sql and added relevant testcases.

A sysadmin login and a non-sysadmin login both see the public role in the sys.server_principals entries.

Example shown below:
```
--BBF Server
1> select name from sys.server_principals
2> go 
name 
------------ 
sysadmin 
rcv
```
## Changes Made:
- Updated `sys.server_principals` view to show `public` role entry
- Added test for the `public` role

### Issues Resolved
Task: BABEL-4518

### Test Scenarios Covered ###
* **Use case based -** yes


* **Boundary conditions -** yes


* **Arbitrary inputs -** yes


* **Negative test cases -** yes


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

Signed-off-by: P Aswini Kumar [aswinii@amazon.com](mailto:aswinii@amazon.com)

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Referencing BABEL-4518 and added suitable tests

shah-nirmit commented on test/JDBC/input/views/sys-server_principals.sql:
There seems to be this file as well as -vu-verify/prepare/cleanup files for server principals lets combine these test files content into -vu-* files

aswiniip replied:
Added the tests in the -vu-verify/prepare/cleanup files

shah-nirmit replied:
I think this is still not addressed , we run the files effectively 2 times once as a single file and then again as vu-prepare/verify/cleanup.

https://github.com/babelfish-for-postgresql/babelfish_extensions/actions/runs/10107002832/job/27950085677?pr=2765#step:11:9215

```
sys-server_principals-vu-prepare:                          Passed! (110/40000ms OK)
sys-server_principals-vu-verify:                           Passed! (1183/40000ms OK)
sys-server_principals-vu-cleanup:                          Passed! (81/40000ms OK)
sys-server_principals:                                     Passed! (387/40000ms OK)

```

Commit:
Added vu-verify tests and include in latest version upgrade script

anju15bharti commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:
```diff
@@ -454,4 +508,4 @@ DROP PROCEDURE sys.babelfish_drop_deprecated_object(varchar, varchar, varchar);
 
 -- Reset search_path to not affect any subsequent scripts
-SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
+SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
unwanted change.

aswiniip replied:
Fixed. Thanks.

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -366,10 +366,33 @@ CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS
+UNION ALL
+SELECT
+CAST(name AS SYS.SYSNAME) AS name,
+CAST(2 AS INT) AS principal_id,
```
Does it exist internally on the PG side?

aswiniip replied:
No it doesn't. This is the following output we get when we query the names in the pg_catalog.pg_roles
```
1> SELECT rolname FROM pg_catalog.pg_roles
2> go
rolname                                                                                                                                                                                                                                                         
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
aswinii                                                                                                                                                                                                                                                         
pg_database_owner                                                                                                                                                                                                                                               
pg_read_all_data                                                                                                                                                                                                                                                
pg_write_all_data                                                                                                                                                                                                                                               
pg_monitor                                                                                                                                                                                                                                                      
pg_read_all_settings                                                                                                                                                                                                                                            
pg_read_all_stats                                                                                                                                                                                                                                               
pg_stat_scan_tables                                                                                                                                                                                                                                             
pg_read_server_files                                                                                                                                                                                                                                            
pg_write_server_files                                                                                                                                                                                                                                           
pg_execute_server_program                                                                                                                                                                                                                                       
pg_signal_backend                                                                                                                                                                                                                                               
pg_checkpoint                                                                                                                                                                                                                                                   
pg_use_reserved_connections                                                                                                                                                                                                                                     
pg_create_subscription                                                                                                                                                                                                                                          
jdbc_user                                                                                                                                                                                                                                                       
bbf_role_admin                                                                                                                                                                                                                                                  
sysadmin                                                                                                                                                                                                                                                        
master_db_owner                                                                                                                                                                                                                                                 
master_dbo                                                                                                                                                                                                                                                      
master_guest                                                                                                                                                                                                                                                    
tempdb_db_owner                                                                                                                                                                                                                                                 
tempdb_dbo                                                                                                                                                                                                                                                      
tempdb_guest                                                                                                                                                                                                                                                    
msdb_db_owner                                                                                                                                                                                                                                                   
msdb_dbo                                                                                                                                                                                                                                                        
msdb_guest                                                                                                                                                                                                                                                      
testlogin                                                                                                                                                                                                                                                       

(28 rows affected)
```

anju15bharti commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -366,10 +366,33 @@ CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS
+UNION ALL
+SELECT
+CAST(name AS SYS.SYSNAME) AS name,
+CAST(2 AS INT) AS principal_id,
```
Not sure if we should keep principal_id as 2 identical to sql server, we should keep this something like -1 or NULL.

aswiniip replied:
Checked with Alex to set the owner to `1` and principal_id to `-1`

anju15bharti commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -366,10 +366,33 @@ CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS
+CAST(1 AS INT) AS owning_principal_id,
+CAST(0 AS sys.BIT) AS is_fixed_role
+FROM (VALUES ('public', 'R')) as dummy_principals(name, type);
 
```
same goes for owning_principal_id.

aswiniip replied:
Checked with Alex to set the owner to `1` and principal_id to `-1`

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -366,10 +366,33 @@ CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS
+CAST(NULL AS SYS.SYSNAME) AS default_database_name,
+CAST(NULL AS SYS.SYSNAME) AS default_language_name,
+CAST(NULL AS INT) AS credential_id,
+CAST(1 AS INT) AS owning_principal_id,
```
Which principal are we referring here as owner? We need to see whether there is any valid metadata exist related to it in other views or not.

HarshLunagariya replied:
For example, In database_principals we are returning -1. Let's clarify this with Stefan/Alex.

aswiniip replied:
Checked with Alex to set the owner to 1

Commit:
Remove unwanted change

Signed-off-by: P Aswini Kumar <aswinii@amazon.com>

Commit:
Changing principal_id to avoid collision

Signed-off-by: P Aswini Kumar <aswinii@amazon.com>

Commit:
updated principal_id in upgrade script

Signed-off-by: P Aswini Kumar <aswinii@amazon.com>

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4518

Commit:
Retrigger tests - Empty commit

Signed-off-by: P Aswini Kumar <aswinii@amazon.com>

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4518

forestkeeper commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:
```diff
@@ -591,6 +591,60 @@ WHERE (SELECT orig_username FROM sys.babelfish_authid_user_ext WHERE rolname = C
+CAST(0 AS sys.BIT) AS is_fixed_role
+FROM (VALUES ('public', 'R')) as dummy_principals(name, type);
+
+GRANT SELECT ON sys.server_principals TO PUBLIC;
```
Why this line ?

aswiniip replied:
Made the changes to directly use values instead of using dummy_principals

forestkeeper commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:
```diff
@@ -591,6 +591,60 @@ WHERE (SELECT orig_username FROM sys.babelfish_authid_user_ext WHERE rolname = C
 
 GRANT SELECT ON sys.user_token TO PUBLIC;
 
+CREATE OR REPLACE VIEW sys.server_principals
```
If we changed the MVU script, then we should add a MVU test for this to verify it works fine with MVU.

anju15bharti replied:
The basic test has been added there in sys-server_principals

forestkeeper reviewed (changes requested):
pls add a test into schedule file to verify the MVU script works fine.

aswiniip commented:
The sys-server_principals test is already present in the MVU script.

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4518

Commit:
Addressed comments

Signed-off-by: P Aswini Kumar <aswinii@amazon.com>

forestkeeper commented on test/JDBC/input/sys-server_principals-vu-verify.mix:
```diff
@@ -15,6 +15,11 @@ FROM sys.server_principals
 
+SELECT name, type, type_desc, default_database_name, default_language_name, credential_id, owning_principal_id, is_fixed_role
+FROM sys.server_principals 
+WHERE name =  'public';
```
Does this test case contain all the babelfish_authid_login_ext.type ?
Like the view sql in case when iterated 'R','U','S','Z' .

forestkeeper reviewed (dismissed):
The code/logic looks good to me, pls merge by Harsh since it should be double check by him whether this logic works well. ( Adding a dummy public to the role , is it possible to create a same name role from customer ? Or is it possible for customer to delete this same role ? Looks like better to have such test cases. )

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4518

Commit:
Show public role entry in sys.server_principals

Currently in babelfish we dont show any 'public' entry. With this change
a public role entry would be shown.

In TSQL, a sysadmin login and a non-sysadmin login both see the public
role in the sys.server_principals entries. In Babelfish, we should show
the public role as well. To achieve this I have added a 'public' entry
in the sys.server_principals by making changes in the ownership.sql and
added relevent testcases.

TASK: BABEL-4518
Signed-off-by: P Aswini Kumar <aswinii@amazon.com>
